<a href="https://colab.research.google.com/github/thejapamarthy/Python-strings-and-if-condition/blob/main/Time_Series_in_Python_(4).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<a href="https://colab.research.google.com/github/afeld/python-public-policy/blob/main/hw_4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **HOMEWORK 4**

# Coding

Goal: Find complaint types that increased or decreased when COVID-19 hit New York Ciy: mid-March 2020.

## Step 0: Setup

For this homework, instead of the data being provided, you will export it directly from the NYC Open Data Portal, as if you were working on your own project.

1. Download the data.
  1. Visit the [311 data](https://data.cityofnewyork.us/Social-Services/311-Service-Requests-from-2010-to-Present/erm2-nwe9/data) page.
  1. From that page, filter the data to `Created Date`s between `01/01/2020 12:00:00 AM` and `03/31/2020 11:59:59 PM`.
  1. It should say "Showing 311 Service Requests 1-100 out of 469,594" near the bottom of the screen.
    - It's ok if the total is slightly different.
  1. Click `Export`.
  1. Click `CSV`. It will start downloading a file.
  1. Rename the file `311_covid.csv`.
1. Upload the data.
  1. Open your NYU [Google Drive](https://drive.google.com/).
  1. [Upload the CSV](https://support.google.com/drive/answer/2424368) to My Drive or a folder under it.
1. Read the data.
  1. Here in Colab, click the folder icon on the left.
  1. Click `Mount Drive`.
  1. Click `Connect to Google Drive`.
  1. Grant permissions.
  1. Under `Files`, you should now see a `drive` folder.
  1. Expand the folder to find your file.
  1. Right-click the file.
  1. Click `Copy path`.

If the above is taking a long time due to have a slow network connection or whatever else, load the data from https://nyu.box.com/shared/static/40srcp74ayro33bxls3z78zs6be9k76g.zip.

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


## Step 1: Load data

Read the data into a DataFrame.

In [ ]:
import pandas as pd

df_2020 = pd.read_csv('https://nyu.box.com/shared/static/40srcp74ayro33bxls3z78zs6be9k76g.zip', low_memory=False)
df_2020

Unique Key  ...                                  Location
0         45289558  ...  (40.677963041857886, -73.98443609121443)
1         45288728  ...   (40.66442190467239, -73.95098201556382)
2         45288240  ...  (40.700366489799876, -73.90543829006366)
3         45287907  ...   (40.59565343138651, -74.00017283917487)
4         45285651  ...   (40.86486556770799, -73.88878325729915)
...            ...  ...                                       ...
469589    45924566  ...  (40.614646565215395, -73.95428903803736)
469590    45921991  ...  (40.682103329312774, -73.80866401575854)
469591    45924709  ...  (40.838796219127325, -73.91813258838923)
469592    45925522  ...   (40.81543236103379, -73.95720386746412)
469593    45921292  ...    (40.85411659342971, -73.8909014364012)

[469594 rows x 41 columns]

## Step 2: Convert dates

Copy code from [Lecture 4](https://colab.research.google.com/github/afeld/python-public-policy/blob/master/lecture_4.ipynb#scrollTo=CHZ-Pqj0bS9w) to convert the `Created Date` to a `datetime`. Print out the types of the columns using [`.info()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.info.html).

In [ ]:
# your code here
df_2020['Created Date'] = pd.to_datetime(df_2020['Created Date'], format='%m/%d/%Y %I:%M:%S %p')
df_2020['Closed Date'] = pd.to_datetime(df_2020['Closed Date'], format='%m/%d/%Y %I:%M:%S %p')
df_2020.info(verbose=True)
df_2020

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 469594 entries, 0 to 469593
Data columns (total 41 columns):
 #   Column                          Non-Null Count   Dtype         
---  ------                          --------------   -----         
 0   Unique Key                      469594 non-null  int64         
 1   Created Date                    469594 non-null  datetime64[ns]
 2   Closed Date                     433908 non-null  datetime64[ns]
 3   Agency                          469594 non-null  object        
 4   Agency Name                     469594 non-null  object        
 5   Complaint Type                  469594 non-null  object        
 6   Descriptor                      461929 non-null  object        
 7   Location Type                   324096 non-null  object        
 8   Incident Zip                    453985 non-null  float64       
 9   Incident Address                436147 non-null  object        
 10  Street Name                     436128 non-null  object 

Unique Key  ...                                  Location
0         45289558  ...  (40.677963041857886, -73.98443609121443)
1         45288728  ...   (40.66442190467239, -73.95098201556382)
2         45288240  ...  (40.700366489799876, -73.90543829006366)
3         45287907  ...   (40.59565343138651, -74.00017283917487)
4         45285651  ...   (40.86486556770799, -73.88878325729915)
...            ...  ...                                       ...
469589    45924566  ...  (40.614646565215395, -73.95428903803736)
469590    45921991  ...  (40.682103329312774, -73.80866401575854)
469591    45924709  ...  (40.838796219127325, -73.91813258838923)
469592    45925522  ...   (40.81543236103379, -73.95720386746412)
469593    45921292  ...    (40.85411659342971, -73.8909014364012)

[469594 rows x 41 columns]

## Step 3: Date counts

Create a DataFrame called `date_counts` that has the count of complaints per Complaint Type per day, then display it.

In [ ]:
# your code here
date_counts=df_2020.groupby('Complaint Type').resample('D',on='Created Date').size().reset_index(name='count_requests')
#date_counts=df_2.resample('D',on='Created Date').size().reset_index(name='count_requests')
date_counts

Complaint Type Created Date  count_requests
0                    APPLIANCE   2020-01-01               6
1                    APPLIANCE   2020-01-02              27
2                    APPLIANCE   2020-01-03              29
3                    APPLIANCE   2020-01-04              11
4                    APPLIANCE   2020-01-05              12
...                        ...          ...             ...
12750             Window Guard   2020-03-22               0
12751             Window Guard   2020-03-23               0
12752             Window Guard   2020-03-24               0
12753             Window Guard   2020-03-25               1
12754  X-Ray Machine/Equipment   2020-01-15               1

[12755 rows x 3 columns]

## Step 4: Plotting over time

Create a line chart of the count of complaints over time, one line per `Complaint Type`.

In [ ]:
# your code here
import plotly.express as px
fig = px.line(date_counts, x="Created Date", y="count_requests", color='Complaint Type')
fig.show()

---

This has the information we need, but is a lot to look at. Let's only show complaint types that changed greatly (in March 2020) relative to the same period in the previous year (March 2019).

## Step 5: March 2020 counts

Create a DataFrame called `mar_counts` that has the count of each `Complaint Type` in March 2020 in a column called `2020`. Use [`.to_frame()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.to_frame.html) (instead of [`.reset_index()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.reset_index.html)) to use the `Complaint Type` as the index. It should end up looking something like this:

Complaint Type | 2020
--- | ---
APPLIANCE | 806
Abandoned Vehicle | 2468
Air Quality | 643
… | …

_Note there is no numeric index._

In [ ]:
# your code here
mar_counts = date_counts[(date_counts['Created Date'].dt.month == 3) & (date_counts['Created Date'].dt.year == 2020)].groupby(by='Complaint Type').sum().rename(columns={'count_requests':'2020'})
mar_counts
#date_counts['2020'] = date_counts.groupby(pd.Grouper(freq='M'))
#df.groupby(pd.Grouper(freq='M'))

#df.groupby(df['date'].dt.strftime('%B'))['Revenue'].sum().sort_values()

2020
Complaint Type                   
APPLIANCE                     806
Abandoned Vehicle            2468
Air Quality                   643
Animal Facility - No Permit     3
Animal in a Park              195
...                           ...
WATER LEAK                   1219
Water Conservation            151
Water Quality                  88
Water System                 2963
Window Guard                    2

[150 rows x 1 columns]

## Step 6: Get March 2019 data

Follow the Setup steps again, but this time:

1. Filter to `Created Date`s between `03/01/2019 12:00:00 AM` and `03/31/2019 11:59:59 PM`.
1. Save as `311_mar_2019.csv`.
1. After uploading to Drive:
   1. In the Colab sidebar, click the Files icon.
   1. Click the Refresh icon.
1. Load that data into a DataFrame called `mar_2019`.
1. Convert the `Created Date` column to a `datetime`.

Similar to Step 0, if having trouble downloading, you can load from https://nyu.box.com/shared/static/ruioqtqfztvhq5q6ng2qedpinrg9bwso.zip.

In [ ]:
# your code here
mar_2019=pd.read_csv('/content/drive/MyDrive/311_mar_2019.csv', low_memory=False)
mar_2019

Unique Key  ...                   Location
0         41842452  ...  (40.7726591, -73.9793233)
1         41845294  ...  (40.6702065, -73.9866873)
2         41845295  ...  (40.8131018, -73.9384814)
3         41845297  ...  (40.7813171, -73.9480578)
4         41845941  ...  (40.8096054, -73.8580546)
...            ...  ...                        ...
222595    42102049  ...  (40.7427407, -73.8850716)
222596    42098653  ...   (40.754584, -73.9991446)
222597    42101731  ...   (40.5760994, -73.997671)
222598    42102543  ...  (40.8478036, -73.9134155)
222599    42098481  ...  (40.8515978, -73.9042654)

[222600 rows x 41 columns]

## Step 7: March 2019 counts

1. Get the `Complaint Type` counts for March 2019.
1. Add these to the `mar_counts` DataFrame as a column called `2019`.

In [ ]:
# your code here
mar_2019['datetime'] = pd.to_datetime(mar_2019['Created Date'], format='%m/%d/%Y %I:%M:%S %p')
mar_2019['Closed Date'] = pd.to_datetime(mar_2019['Closed Date'], format='%m/%d/%Y %I:%M:%S %p')
date_counts2019=mar_2019.groupby('Complaint Type').resample('D',on='datetime').size().reset_index(name='count_requests')
mar2019counts = date_counts2019[(date_counts2019['datetime'].dt.month == 3) & (date_counts2019['datetime'].dt.year == 2019)].groupby(by='Complaint Type').sum().rename(columns={'count_requests':'2019'})
mar2019counts
mar_counts = pd.concat([mar_counts,mar2019counts], axis=1, join='inner')
mar_counts

2020  2019
Complaint Type                         
APPLIANCE                     806  1042
Abandoned Vehicle            2468     1
Air Quality                   643   642
Animal Facility - No Permit     3    10
Animal in a Park              195   211
...                           ...   ...
WATER LEAK                   1219  2603
Water Conservation            151   298
Water Quality                  88   108
Water System                 2963  3880
Window Guard                    2     1

[138 rows x 2 columns]

## Step 8: Percent change

Use `mar_counts` to calculate the percent change from March 2019 to March 2020 for each `Complaint Type`. Save as the `pct_change` column. Reminder that you can do arithmetic between columns just like you would between variables or even numbers. Should result in something like this:

Complaint Type | 2020 | 2019 | pct_change
--- | --- | --- | ---
APPLIANCE | 806 | 1042 | -0.22
Abandoned Vehicle | 2468 | 1 | 2467.00
Air Quality | 643 | 642 | 0.00
… | … | … | …

In [ ]:
# your code here
mar_counts['pct_change']= ((mar_counts['2020']-mar_counts['2019'])/mar_counts['2019'])
mar_counts

2020  2019   pct_change
Complaint Type                                      
APPLIANCE                     806  1042    -0.226488
Abandoned Vehicle            2468     1  2467.000000
Air Quality                   643   642     0.001558
Animal Facility - No Permit     3    10    -0.700000
Animal in a Park              195   211    -0.075829
...                           ...   ...          ...
WATER LEAK                   1219  2603    -0.531694
Water Conservation            151   298    -0.493289
Water Quality                  88   108    -0.185185
Water System                 2963  3880    -0.236340
Window Guard                    2     1     1.000000

[138 rows x 3 columns]

## Step 9: Filter

Filter to `Complaint Type`s that both:

- Occurred at least 50 times in March 2020
- Changed (increased _or_ decreased) by more than 90%

and save the DataFrame as `top_changed`. A couple of things that may be helpful:

- [Selecting Subsets of Data in Pandas](https://medium.com/dunder-data/selecting-subsets-of-data-in-pandas-39e811c81a0c#0eb4), starting from "Multiple condition expression"
- [Getting absolute values](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.abs.html)

In [ ]:
# your code here
top_changed= mar_counts.loc[(mar_counts["2020"]>50) & abs((mar_counts["pct_change"])>0.9)]
top_changed


2020  2019   pct_change
Complaint Type                                      
Abandoned Vehicle            2468     1  2467.000000
Consumer Complaint           9324  1228     6.592834
Drug Activity                 267   134     0.992537
Homeless Person Assistance   1475   728     1.026099
Noise - Helicopter            368    96     2.833333
Non-Emergency Police Matter  1846   751     1.458056
Urinating in Public            63    33     0.909091

## Step 10: Top changed

Filter the `date_counts` to only the `top_changed` `Complaint Type`s. Save as `top_changed_by_day`.

In [ ]:
# your code here
#date_counts=df_2020.groupby('Complaint Type').resample('D',on='Created Date').size().reset_index(name='count_requests')
top_changed_by_day= pd.merge(date_counts, top_changed, on=['Complaint Type'], how='inner')
top_changed_by_day=top_changed_by_day.drop(columns=['2020','2019','pct_change'])
#date_counts[date_counts.set_index(['Complaint Type']).index.isin(top_changed.set_index(['Complaint Type']).index)]
top_changed_by_day


Complaint Type Created Date  count_requests
0      Abandoned Vehicle   2020-01-01              63
1      Abandoned Vehicle   2020-01-02             156
2      Abandoned Vehicle   2020-01-03             151
3      Abandoned Vehicle   2020-01-04              82
4      Abandoned Vehicle   2020-01-05              97
..                   ...          ...             ...
626  Urinating in Public   2020-03-26               2
627  Urinating in Public   2020-03-27               0
628  Urinating in Public   2020-03-28               2
629  Urinating in Public   2020-03-29               4
630  Urinating in Public   2020-03-30               5

[631 rows x 3 columns]

## Step 11: Plotting changed complaints

Make a similar plot to Step 4, but with only the top complaints (`top_changed_by_day`).

In [ ]:
# your code here
import plotly.express as px
fig = px.line(top_changed_by_day, x="Created Date", y="count_requests", color='Complaint Type')
fig.show()

## Question 0

***Did the change of any of the `Complaint Type`s in Step 10/11 surprise you? Why or why not? (Speak at least one specifically.)***
Urinating in public complaint surprised me because Mar 2020 techically had lockdown which means we should have had fewer complaints on public urination

YOUR RESPONSE HERE

Then, give these a read:

- [NY Daily News article](https://www.nydailynews.com/coronavirus/ny-coronavirus-price-gouging-new-york-city-20200429-z5zs4ygfxbcmrpgzfrnlbxsnea-story.html)
- [Press release from Department of Consumer and Worker Protection](https://www1.nyc.gov/site/dca/media/pr031720-DCWP-Emergency-Rule-Price-Gouging-Illegal.page)

Overall caveat for this assignment: [**correlation does not imply causation**](https://www.khanacademy.org/math/probability/scatterplots-a1/creating-interpreting-scatterplots/v/correlation-and-causality).

## Question 1

***Did you work with anyone else on this assignment?***

YOUR RESPONSE HERE yes for Step 7

## Bonus: Charting against COVID-19 case counts

Let's take a look at the `Consumer Complaint`s against the COVID-19 case numbers in NYC in the same graph. You'll need to:

1. Find data that provides the COVID-19 case counts for NYC by day.
1. Create a DataFrame with only the `Consumer Complaint` `Complaint Type` counts, by day.
1. Chart the two against each other for February through March.

The result should look something like this:

![bonus solution chart](https://github.com/afeld/python-public-policy/blob/main/img/hw_4_bonus.png?raw=1)

Some resources that may be helpful, minus the blacked-out part:

- [Reading CSV data from GitHub](https://projectosyo.wixsite.com/datadoubleconfirm/single-post/2019/04/15/Reading-csv-data-from-Github---Python)
- [Two Y Axes in plotly](https://plotly.com/python/multiple-axes/#two-y-axes)
   - Note that the `plotly.graph_objects` syntax is a bit different than the `plotly.express` syntax we've been using. With `go.Scatter()`, you don't provide the DataFrame and the names of the columns; you pass `x` and `y` as lists/Series of the values themselves.
- [Setting the Range of Axes Manually in plotly](https://plotly.com/python/axes/#setting-the-range-of-axes-manually)

In [ ]:
# your code here
covid19=pd.read_csv('/content/drive/MyDrive/COVID-19_counts.csv', low_memory=False)
covid19
consumercomplaint=top_changed_by_day.loc[top_changed_by_day["Complaint Type"] =="Consumer Complaint"]
consumercomplaint
import plotly.graph_objects as go
from plotly.subplots import make_subplots
fig = go.Figure(go.Scatter(
    mode = "lines+markers",
    y = [1000,2000,3000,4000,5000,6000],
    x = ["Feb 2 2020", "Feb 16", "Mar 1", "Mar 15", "Mar 29"]))
fig = make_subplots(specs=[[{"secondary_y": True}]])
fig.add_trace(
    go.Scatter(covid19,x=["Feb 2 2020", "Feb 16", "Mar 1", "Mar 15", "Mar 29"], y=[1000,2000,3000,4000,5000,6000],name="Case Count"),
    secondary_y=False,
)
fig.add_trace(
    go.Scatter(consumercomplaint,x=["Feb 2 2020", "Feb 16", "Mar 1", "Mar 15", "Mar 29"], y=[1000,2000,3000,4000,5000,6000], name="Consumer Complaints"),
    secondary_y=True,
)
fig.update_layout(
    title_text="Consumer Complaints in 311 vs Covid Cases in NYC"
)
fig.update_xaxes(title_text="Date")
fig.update_yaxes(title_text="<b>primary</b> Case Count", secondary_y=False)
fig.update_yaxes(title_text="<b>secondary</b> Case Count", secondary_y=True)

fig.show()


ValueError: ignored

### Bonus Question 1: What observations do you have?

YOUR RESPONSE HERE

# Tutorial

**Read [spaCy 101](https://spacy.io/usage/spacy-101#whats-spacy) through `Linguistic annotations`** (stop at `Pipelines`). It's ok if it doesn't all make sense—this is mostly to introduce you to some terminology for the next topic: natural language processing!